src/Rules/Comparison/ImpossibleCheckTypeHelper.php

```diff
@@ -147,7 +147,7 @@ public function findSpecifiedType(
 							foreach ($haystackArrayTypes as $haystackArrayType) {
 								if ($haystackArrayType instanceof ConstantArrayType) {
 									foreach ($haystackArrayType->getValueTypes() as $i => $haystackArrayValueType) {
-										if ($haystackArrayType->isOptionalKey($i)) {
+										if ($haystackArrayValueType instanceof UnionType || $haystackArrayType->isOptionalKey($i)) {
```

Comment:
I believe there's a better way to ask this. Something like `count($haystackArrayValueType->getFiniteTypes()) === 1`.